bob/src/cluster/quorum.rs

```diff
@@ -272,18 +272,64 @@ impl Cluster for Quorum {
     }
 
     async fn exist(&self, keys: &[BobKey]) -> Result<Vec<bool>, Error> {
-        let keys_by_nodes = group_keys_by_nodes(&self.mapper, keys);
-        debug!(
-            "EXIST Nodes for fan out: {:?}",
-            &keys_by_nodes.keys().flatten().collect::<Vec<_>>()
-        );
         let len = keys.len();
         let mut exist = vec![false; len];
-        for (nodes, (keys, indexes)) in keys_by_nodes {
-            let cluster_results = LinkManager::exist_on_nodes(&nodes, &keys).await;
-            for result in cluster_results.into_iter().flatten() {
-                for (&r, &ind) in result.inner().iter().zip(&indexes) {
-                    exist[ind] |= r;
+
+        // filter local keys
+        let (indices, local_keys) = filter_local_keys(keys, &self.mapper);
+        // end
+        debug!("local keys {:?}", local_keys);
+        debug!("local indices {:?}", indices);
+        if local_keys.len() > 0 {
+            let result = exist_on_local_node(&self.backend, &local_keys).await?;
+            for (idx, r) in indices.into_iter().zip(result.into_iter()) {
+                exist[idx] |= r;
+            }
+            debug!("exist after local node {:?}", exist);
+        }
+
+        // filter keys that were not found
+        let local_alien = filter_not_found(&exist, keys);
+        // end
+        debug!("local alien keys {:?}", local_alien);
+        if local_alien.len() > 0 {
+            let result = exist_on_local_alien(&self.backend, &local_alien).await?;
+            update_exist(&mut exist, &result);
+            debug!("exist after local alien {:?}", exist);
+        }
+
+        // filter remote not found keys by nodes
+        let remote_keys = filter_remote_not_found_by_nodes(&exist, keys, &self.mapper);
+        // end
+
+        debug!("remote keys by nodes {:?}", remote_keys);
+        if remote_keys.len() > 0 {
+            for (nodes, (keys, indices)) in remote_keys {
+                let result = exist_on_remote_nodes(&nodes, &keys).await;
+                for res in result.into_iter().flatten() {
+                    for (&r, &ind) in res.inner().iter().zip(&indices) {
+                        exist[ind] |= r;
+                    }
+                }
+            }
+            debug!("exist after remote nodes {:?}", exist);
+        }
+
+        // filter remote not found keys
+        let remote_alien = filter_not_found(&exist, keys);
+        // end
+        if remote_alien.len() > 0 {
+            // filter remote nodes
+            let remote_nodes = filter_remote_nodes(&self.mapper);
```

Comment:
This will lead to skipping local aliens for keys, that don't belong to the local node